### PR TITLE
i#3574 binutils decoder/encoder tests: Add binutils decode/encode tests to DynamoRIO.

### PR DESCRIPTION
**Pull request by hgreving2304**
This patch adds a test suite based on binutil's gas test suite. It is not equivalent to
binutil's suite. For example, DynamoRio does not have an assembler reading mnemonics.

The suite adds a collection of decoder/encoder tests and compares its disassembler output.
The expected disassembler strings have been generated by DynamoRIO at the time this suite
was created. Hence, it is not golden reference with respect to binutils. There are minor
differences between objdump and DynamoRIO's disassembler output that would need to get
resolved. This is a possible future improvement to this suite.

Decoding and encoding errors have been debugged and inspected (xref #3577, #3578, #3581).
Unsupported instructions are either referenced by these issues, or have been removed
from the original binutils test suite. For example, unsupported CPL 0 instructions have
been removed.

The suite was moved to third_party to honor GNU GPL. Only the .asm files of this suite
are subject to GNU GPL.

Large parts of the regression are currently commented out, specifically AVX-512 (xref
i#1312), MPX, SHA, VMX (xref i#3581) and others (issues mentioned above).

Issue: #3574

### CODE REVIEW

**Comment by derekbruening** [comment]
> Large parts of the regression are currently commented out

Commented out code is disallowed in our code base.  Please s/commented out/disabled/.

**Comment by derekbruening** [comment]
I see third_party/test_decenc -- I would expect the project name as the top-level dir, so third_party/binutils/test_decenc.

**Comment by hgreving2304** [comment]
> > Large parts of the regression are currently commented out
> 
> Commented out code is disallowed in our code base. Please s/commented out/disabled/.

It's actually just test names in a list in suite/tests/CMakeLists.txt. I would prefer having it there so it is obvious what's missing. Otherwise I'll remove and add to the README file.

**Comment by derekbruening** [comment]
> > > Large parts of the regression are currently commented out
> >  
> > Commented out code is disallowed in our code base. Please s/commented out/disabled/.
> 
> It's actually just test names in a list in suite/tests/CMakeLists.txt. I would prefer having it there so it is obvious what's missing. Otherwise I'll remove and add to the README file.

There is no request to remove content: simply to disable in a way that satisfies the style guide.

**Comment by hgreving2304** [comment]
> > > > Large parts of the regression are currently commented out
> > > 
> > > 
> > > Commented out code is disallowed in our code base. Please s/commented out/disabled/.
> > 
> > 
> > It's actually just test names in a list in suite/tests/CMakeLists.txt. I would prefer having it there so it is obvious what's missing. Otherwise I'll remove and add to the README file.
> 
> There is no request to remove content: simply to disable in a way that satisfies the style guide.

Will update.

File: api/samples/stl_test.cpp [hunk 1]
```diff
@@ -71,6 +71,7 @@ dr_init(client_id_t client_id)
 {
     int i;
     bool success = true;
+    (void)success;
```
Comment (derekbruening) [inline api/samples/stl_test.cpp]: This seems to have no relation to the rest of the PR: a mistake?  Supposed to be in another PR I assume.

File: suite/runsuite_common_pre.cmake [hunk 2]
```diff
@@ -347,9 +347,9 @@ elseif (arg_use_make)
   if (have_cygwin)
     # seeing errors building in parallel: pdb collision?
     # can't repro w/ VERBOSE=1
-    set(CTEST_BUILD_COMMAND_BASE "${MAKE_COMMAND} -j2")
+    set(CTEST_BUILD_COMMAND_BASE "${MAKE_COMMAND} -j2 VERBOSE=1")
```
Comment (derekbruening) [inline suite/runsuite_common_pre.cmake]: I'm seeing multiple changes that presumably should be temporary but are not marked that way!  Is this ready for review?  Please look through it.

Comment (hgreving2304) [reply suite/runsuite_common_pre.cmake]: Do not review at this time, wait for PTAL (debugging clang error).

File: api/samples/stl_test.cpp (hunk 1 above)
Comment (hgreving2304) [reply api/samples/stl_test.cpp]: Also temporary change, will get removed (debugging clang error).

Comment (derekbruening) [reply api/samples/stl_test.cpp]: I would suggest following the habit of always marking temp changes as such with comments, esp in shared branches.  Multiple benefits: easier to remember what needs to be reverted once done debugging; easier for someone else to pick up the PR if necessary.

Comment (hgreving2304) [reply api/samples/stl_test.cpp]: Ack/Done.

Comment (derekbruening) [reply api/samples/stl_test.cpp]: This is what `NOCHECKIN` was meant for, and we have checks to ensure a line which contains `NOCHECKIN` (usually at the end `// NOCHECKIN`) does not get merged.  For testing on Travis though it will fail that check up front.

Comment (derekbruening) [reply api/samples/stl_test.cpp]: Quoting from https://github.com/DynamoRIO/dynamorio/wiki/Code-Style:

```
Mark any temporary or unfinished code unsuitable for committing with a NOCHECKIN comment. The make/codereview.cmake script will remind you to clean up the code.

 x = 4; /* NOCHECKIN: temporary debugging change */
```

**Comment by derekbruening** [comment]
> I think the style guide is w.r.t. C-code. In this case, it's a cmake file and a proper comment was added. I am not sure if it becomes more readable if I am adding a conditional around if and break up the list. Might be better to remove it then and add to README file. I would actually leave it as it is right now.

It should apply to all code.  A conditional seems better.  As mentioned before please also update the commit message.

**Comment by derekbruening** [comment]
Is this adding a whole bunch of separately launched tests, or are these combined into one?  If the former, does it add noticeable time to the suite (esp on Windows) to launch all these processes?

**Comment by hgreving2304** [comment]
> s this adding a whole bunch of separately launched tests, or are these combined into one? If the former, does it add noticeable time to the suite (esp on Windows) to launch all these processes?

Yes, it's individual tests for ctest. I did think about encapsulating them in something like BUILD_DECENC_TESTS=ON|OFF, but they compile very quick and also run very quick, I think there's no need to separate them out.

**Comment by hgreving2304** [comment]
run arm tests

**Comment by hgreving2304** [comment]
> > I think the style guide is w.r.t. C-code. In this case, it's a cmake file and a proper comment was added. I am not sure if it becomes more readable if I am adding a conditional around if and break up the list. Might be better to remove it then and add to README file. I would actually leave it as it is right now.
> 
> It should apply to all code. A conditional seems better. As mentioned before please also update the commit message.

Sure. Done (do not review patch yet).

File: api/samples/stl_test.cpp (hunk 1 above)
Comment (hgreving2304) [reply api/samples/stl_test.cpp]: Done.

**Comment by hgreving2304** [comment]
PTAL.

File: third_party/binutils/test_decenc/drdecode_decenc_x86_test_absrel_s.runcmp [hunk 3]
```diff
@@ -0,0 +1 @@
+presence of this file indicates to use runcmp.cmake
```
Comment (derekbruening) [inline third_party/binutils/test_decenc/drdecode_decenc_x86_test_absrel_s.runcmp]: The alternative method of just defining a variable (as I originally suggested over chat) seems simpler and cleaner and avoids all of these files.

File: third_party/binutils/test_decenc/drdecode_decenc_x86.c [hunk 4]
```diff
@@ -0,0 +1,3786 @@
...
+
+/* Arbitrary pc for pc-relative operands for consistent output */
+#    define ORIG_PC ((app_pc)0x10000000)
+
+static void
+test_s(byte *subtest_asm)
+{
+    byte *orig_pc;
+    instr_t *instr = instr_create(GD);
+    byte buf[16];
```
Comment (derekbruening) [inline third_party/binutils/test_decenc/drdecode_decenc_x86.c]: 16 is not long enough.  17 is the classic max length for x86 instrs: see MAX_INSTR_LENGTH in DR.

File: third_party/binutils/test_decenc/drdecode_decenc_x86.c [hunk 5]
```diff
@@ -0,0 +1,3786 @@
...
+    byte *dec_pc = (byte *)subtest_asm;
```
Comment (derekbruening) [inline third_party/binutils/test_decenc/drdecode_decenc_x86.c]: Redundant cast: remove.

File: third_party/binutils/test_decenc/drdecode_decenc_x86.c [hunk 6]
```diff
@@ -0,0 +1,3786 @@
...
+    while (*(uint64 *)dec_pc != 0x0102030405060708) {
+        instr_reset(GD, instr);
+        orig_pc = dec_pc;
+        decode_from_copy(GD, dec_pc, ORIG_PC + (dec_pc - start), instr);
+        dec_pc = disassemble_from_copy(GD, dec_pc, ORIG_PC + (dec_pc - start), STDOUT,
+                                       false, true);
+        instr_set_raw_bits_valid(instr, false);
+        instr_encode_to_copy(GD, instr, buf, ORIG_PC + (dec_pc - start));
+    }
+}
```
Comment (derekbruening) [inline third_party/binutils/test_decenc/drdecode_decenc_x86.c]: bug: memory leak: need to free instr

File: third_party/binutils/test_decenc/drdecode_decenc_x86.c [hunk 7]
```diff
@@ -0,0 +1,3786 @@
...
+    byte buf[16];
+    byte *dec_pc = (byte *)subtest_asm;
+    byte *start = dec_pc;
+    int count = 0;
+    disassemble_set_syntax(DR_DISASM_ATT);
+    while (*(uint64 *)dec_pc != 0x0102030405060708) {
+        instr_reset(GD, instr);
+        orig_pc = dec_pc;
+        decode_from_copy(GD, dec_pc, ORIG_PC + (dec_pc - start), instr);
+        dec_pc = disassemble_from_copy(GD, dec_pc, ORIG_PC + (dec_pc - start), STDOUT,
```
Comment (derekbruening) [inline third_party/binutils/test_decenc/drdecode_decenc_x86.c]: The order makes this look like inefficient code: if you've already decoded the instr you'd use instr_disassemble() instead of re-decoding it by using a raw disassemble.  I would swap and do the disassemble first: the decode belongs with the two lines below.

File: third_party/binutils/test_decenc/drdecode_decenc_x86.c [hunk 8]
```diff
@@ -0,0 +1,3786 @@
...
+    byte *orig_pc;
+    instr_t *instr = instr_create(GD);
+    byte buf[16];
+    byte *dec_pc = (byte *)subtest_asm;
+    byte *start = dec_pc;
+    int count = 0;
+    disassemble_set_syntax(DR_DISASM_ATT);
+    while (*(uint64 *)dec_pc != 0x0102030405060708) {
+        instr_reset(GD, instr);
+        orig_pc = dec_pc;
```
Comment (derekbruening) [inline third_party/binutils/test_decenc/drdecode_decenc_x86.c]: This variable is never used: delete.

Also: IMHO it is bad practice to "case-shadow" variables with `orig_pc` and `ORIG_PC`.

File: third_party/binutils/test_decenc/drdecode_decenc_x86.c [hunk 9]
```diff
@@ -0,0 +1,3786 @@
...
+#        endif
+#        ifdef test_xsave_s
+    dr_printf("test_xsave_s:\n");
+    test_s((byte *)test_xsave_s_asm);
+#        endif
+#        ifdef test_xsavec_s
+    dr_printf("test_xsavec_s:\n");
+    test_s((byte *)test_xsavec_s_asm);
+#        endif
+#
```
Comment (derekbruening) [inline third_party/binutils/test_decenc/drdecode_decenc_x86.c]: This approach seems to require a lot of work to edit.
If I want to add or remove a test, I have to go and edit 7 different places in shared files:

1. The list in suite/tests/CMakeLists.txt.
1. The function declaration at the top of this file.
1. The ifdef around the print and test_s call.
1. The print of the test name.
1. The call to test_s().
1. Another ifdef down here.
1. The include of the .asm file.

That is a lot of places storing the unique name, for each of 400 separate names.

I then have to create 3 new files: the .asm, the .expect, and the .runcmp.  The .runcmp can easily be eliminated, but that is a lot of places where these names are stored.

I only looked at a couple .asm files, but many seem pretty short.  In fact, they are not that much longer than the 3 lines here for including them.  My proposal is to inline all the .asm files here and have a single .expect file.  That eliminates all of the duplication of these separate test names and IMHO makes it simpler both to read the code and to maintain it.  We can have comments labeling the test boundaries in the expect file: so we'd have a subtest name once here in the code and once in the expect file.

I don't see a big benefit from having these as completely separate tests: only in seeing what failed, but that is a rare occurrence.  The downsides outweigh that IMHO: it is awkward dealing with 400 names all duplicated in various places, and 1200 (!) separate files, and building and running 400 separate binaries takes a lot longer than building just one.

I prefer fewer files when we're talking about hundreds or thousands of them, but an alternative could be separate code and expect files, and just get rid of all the declarations and ifdefs and concatenate everything together here and in the single master expect file.

File: third_party/binutils/test_decenc/drdecode_decenc_x86_test_absrel_s.runcmp (hunk 3 above)
Comment (hgreving2304) [reply third_party/binutils/test_decenc/drdecode_decenc_x86_test_absrel_s.runcmp]: Done.

File: third_party/binutils/test_decenc/drdecode_decenc_x86.c (hunk 4 above)
Comment (hgreving2304) [reply third_party/binutils/test_decenc/drdecode_decenc_x86.c]: I actually thought it's 15!

File: third_party/binutils/test_decenc/drdecode_decenc_x86.c (hunk 5 above)
Comment (hgreving2304) [reply third_party/binutils/test_decenc/drdecode_decenc_x86.c]: Done.

File: third_party/binutils/test_decenc/drdecode_decenc_x86.c (hunk 6 above)
Comment (hgreving2304) [reply third_party/binutils/test_decenc/drdecode_decenc_x86.c]: Added instr_destroy at the end of the test.

File: third_party/binutils/test_decenc/drdecode_decenc_x86.c (hunk 8 above)
Comment (hgreving2304) [reply third_party/binutils/test_decenc/drdecode_decenc_x86.c]: Done.

File: third_party/binutils/test_decenc/drdecode_decenc_x86.c (hunk 7 above)
Comment (hgreving2304) [reply third_party/binutils/test_decenc/drdecode_decenc_x86.c]: Done.

File: third_party/binutils/test_decenc/drdecode_decenc_x86.c (hunk 9 above)
Comment (hgreving2304) [reply third_party/binutils/test_decenc/drdecode_decenc_x86.c]: I wanted to keep the original separation from binutils, such that if they add something, one can easily see what they've added. Yes, you have to edit a few places, but this is just a one time thing that (will-) almost never happen(s). Re-factoring this like this will take me another few hours potentially. The advantage is also to be able to easily disable the parts that are currently not working, just by editing them in suite/tests/CMakeLists.txt (#1312, #3577, #3578, #3581).

Comment (derekbruening) [reply third_party/binutils/test_decenc/drdecode_decenc_x86.c]: > I wanted to keep the original separation from binutils, such that if they add something, one can easily see what they've added.

I don't see how the organization in our repo relates to that: simply look at the binutils version control.  The names would still be in comments in any case which is just as easy to check as a list of files.

If these were straight from binutils then I could see the argument: but these are heavily edited.  Or if they were truly separate: but they are already bundled and #included into a single .c file anyway.  Once we have that, why build and run as separate tests?  That part seems to not offer advantages.  Agreed?

The separate files is then a further issue that perhaps should go to a wider audience then if we disagree.

> Yes, you have to edit a few places, but this is just a one time thing that (will-) almost never happen(s). Re-factoring this like this will take me another few hours potentially. The advantage is also to be able to easily disable the parts that are currently not working, just by editing them in suite/tests/CMakeLists.txt (#1312, #3577, #3578, #3581).

Disabling is going to be rarer than reading, adding, editing.

File: third_party/binutils/test_decenc/addr16.asm [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+#undef FUNCNAME
+#define FUNCNAME test_addr16_s_asm
+DECLARE_FUNC_SEH(FUNCNAME)
+GLOBAL_LABEL(FUNCNAME:)
+        /* There is intentionally no prologue and epilogue. This function is not
+         * supposed to be actually called. We are using the function like a label
+         * for the decoder.
+         */
+        END_PROLOG
+        RAW(67) RAW(a0) RAW(98) RAW(08) RAW(67) RAW(66)
```
Comment (derekbruening) [inline third_party/binutils/test_decenc/addr16.asm]: So I went and looked at my (old I guess, 2.18.50) binutils sources I happen to have sitting around, and there I see:
```
 $ cat gas/testsuite/gas/i386/addr16.s
	.text
	addr16 mov	0x0898,%al
	addr16 mov	0x0898,%ax
	addr16 mov	0x0898,%eax
	addr16 mov	%al,0x0898
	addr16 mov	%ax,0x0898
	addr16 mov	%eax,0x0898
```

That is readable.  Why were those converted into raw bytes which no human can read?  This does not seem conducive to code reading or understanding, to updating this file if binutils augments or changes their version, etc.

Comment (derekbruening) [reply third_party/binutils/test_decenc/addr16.asm]: I understand why you might convert to raw bytes for avx-512 or other recent ISA features where a test machine may not be able to assemble if in assembler form, but not for this test.

An alternative to assembling at test time would be to take the *.s files and assemble them yourself once into a binary using a checked-in script that is run with a recent assembler, so we don't need *any* raw entries and can use the binutils .s files as-is.  The test then decodes from a public symbol in a checked-in binary.  Maybe it's too complex and we'd only do this if we expected to updated from binutils frequently and really wanted to have unchanged .s files from them.

Comment (hgreving2304) [reply third_party/binutils/test_decenc/addr16.asm]: The raw bytes come from the .d results files in the binutils testsuite. The binutils suite assembles first, we don't have an assembler that reads at&t or any other syntax.

I don't see that we ever have to change these files even if binutils changes. It's all valid x86 encodings.

Comment (derekbruening) [reply third_party/binutils/test_decenc/addr16.asm]: > we don't have an assembler that reads at&t or any other syntax.

?  Sure we do: we have a number of assembly files which we have to assemble as part of building DR and at least on Linux AT&T is perfectly supported by the toolchain we rely on.

Comment (hgreving2304) [reply third_party/binutils/test_decenc/addr16.asm]: You mean if we use the local toolchain. Sure. That's another way how to do it. I (wanted to) made (make) it independent from using an assembler.

Comment (derekbruening) [reply third_party/binutils/test_decenc/addr16.asm]: > You mean if we use the local toolchain. Sure. That's another way how to do it. I (wanted to) made (make) it independent from using an assembler.

But it is already using an assembler, and the local toolchain: the local assembler is taking these RAW declarations and creating binary code.  Is the reason for using RAW to make it build on Windows where it was easier to use RAW than to convert the AT&T into Intel?

Comment (hgreving2304) [reply third_party/binutils/test_decenc/addr16.asm]: As a side note, I think we're testing our decoder/encoder one way or the other.

Windows was not the reason. But that's a good reason, too. I think the gas test suite is testing something different than us. They are testing the gas assembler, i.e. taking mnemonics and corner cases of mnemonics and encoding them. Then they cross check the results with objdump. For DynamoRIO, we're testing the raw-bytes decoder (and the encoder kind of by making a round-trip). So I looked at the binutils test suite just as a source to get interesting x86 encoding bytes.

The fact that they're testing gas includes invalid cases. If you took the .s files, you had to filter out the _invalid_ cases first. That's in addition to the _valid_ cases we do not support (e.g. CPL-0 instructions). I already had to filter the encodings for all the cases we either don't support or not yet support (i#3577, i#3578, i#1312, i#3581). I didn't want to filter the mnemonics as well, that don't add much except, yes, readability.

Comment (derekbruening) [reply third_party/binutils/test_decenc/addr16.asm]: The DR decoder does support CPL-0 instructions.

Comment (derekbruening) [reply third_party/binutils/test_decenc/addr16.asm]: Readability is a big deal.   Making these just a sequence of hex numbers makes them unreadable and disconnects them from the original files.  If they're just piles of hex numbers disconnected from the binutils files I would much prefer to have them in one file instead of dealing with 400 (!) separate tiny unreadable files.

Comment (hgreving2304) [reply third_party/binutils/test_decenc/addr16.asm]: Not sure what you mean by "just piles of hex numbers". They could be "just piles of mnemonics" then as well. How do you want to handle the currently disabled subtests then? Right now they are conveniently listed in suite/tests/CMakeLists.txt. If I pipe them into one big file with comments so that I can see which ones were which, do you want to ifdef out the parts that are currently disabled per DISABLED_UNTIL_BUG_<N>?

Comment (hgreving2304) [reply third_party/binutils/test_decenc/addr16.asm]: > The DR decoder does support CPL-0 instructions.

Ok, I did see some that the decoder does not have, I assumed we didn't generally support ring 0 instructions. I will add them to DDK-3578 (for example clac, slac).

Comment (derekbruening) [reply third_party/binutils/test_decenc/addr16.asm]: @Carrotman42 any opinion on having these 400 separated raw-hex files each as a separate test using ifdefs in one combined file, vs inlining into that file for a single test?

Comment (hgreving2304) [reply third_party/binutils/test_decenc/addr16.asm]: I don't have a problem with converting it to a one file setup (just with re-writing everything using .s files). Done.

File: third_party/binutils/test_decenc/drdecode_decenc_x86.c (hunk 9 above)
Comment (hgreving2304) [reply third_party/binutils/test_decenc/drdecode_decenc_x86.c]: See below.
